llvm/lib/Analysis/ScalarEvolution.cpp

```diff
@@ -1793,6 +1793,18 @@ const SCEV *ScalarEvolution::getZeroExtendExprImpl(const SCEV *Op, Type *Ty,
       return getAddExpr(Ops, SCEV::FlagNUW, Depth + 1);
     }
 
+    const SCEVConstant *C;
+    const SCEV *A;
+    // zext (C + A)<nsw> -> (sext(C) + zext(A))<nsw> if zext (C + A)<nsw> >=s 0
+    // and A >=s V.
```

Comment:
```suggestion
    // and A >=s C.
```

Reply:
Though, wouldn't the more natural fold here be something like https://alive2.llvm.org/ce/z/RF9XaY? For the case where A >= 0 the sext would become a zext.

Reply:
Here are some changes with the generalization: https://github.com/llvm/llvm-project/pull/142599/commits/7f8f9370d9849921580d19c09ec1ff91b1351b65

* in `@add_nsw_zext_fold_results_in_sext` we have a in more complex expansion
* in  `@fold_add_zext_to_sext` we miss some re-use during expansion.

Reply:
Sounds good, so should we go with the current version then straight away or first land the restricted version which is less likely to intoduce any regressions?